### ISSUE TITLE
ClientInterceptor per call parameters

### ISSUE THREAD

**Issue by elandau** [grpc/grpc-java#1800]
Our generated clients come pre-configured with a set of approximately 10 interceptors covering functionality from response caching to metrics reporting.  We'd like our users to be able to customize these interceptors per call.  What is the recommended mechanism to pass per call parameters to a client already configured with ClientInterceptors?  

Here's an example for setting up a client

```java
HelloStub client = HelloGrpc.newBlockingStub(
    ClientInterceptors.intercept(NettyChannelBuilder.forAddress("localhost:9000").build(),
        new ResponseCacheInterceptor()),
        new MetricsInterceptor()
    );
```

When using the client we'd like our users to do something like,

```java
client.withOption(ResponseCacheInterceptor.CACHE_KEY, "foo")
    .sayHello(request);
```

But CallOptions currently does not support custom options.  Is something like this in the works?

Alternatively, we could use the Context but it seems overkill to go through a thread local for this.

```java
Context.current().withValue(ResponseCacheInterceptor.CACHE_KEY, "foo").call(()-> {
    client.sayHello(request);
});
```

Any other recommendations on how this could be accomplished?

**Comment by ejona86** [comment]
Just for completeness, how bad is it for your environment to add the cache interceptor directly to the stub:
```java
client.withInterceptors(cache.newInterceptor("foo"))
    .sayHello(request);
```

It does mean more pain with sharing the cache. @louiscryan, thoughts?

I think we've been thinking up until this point that you would either create the interceptor per-call or use `Context`. But adding custom options to `CallOptions` does appear more convenient. It'd be nice if every object didn't have a map (LB Attributes, Metadata, Context, Transport Attributes), but that tendency is already a well-known phenomenon.

I'll note that caching _may_ be a pathological case since it needs to share state across calls _and_ be configured per-call. That doesn't seem unreasonable, but I also don't have a good idea on how often that would occur. I'm also not sure if the other languages could easily cope with the same situation.

**Comment by elandau** [comment]
The issue with installing per call interceptors is that we already install about 10 interceptors and order matters a lot.  Processing the cache first would cause things to happen out of expected sequence.

Regarding caching specifically we actually allow several levels of caching: distributed, system and per request context.  The complexity here is part of the reason why we want caching to be configured on the client as opposed to per call.  Note also that caching is an optimization that we have currently implemented in Java and plan to port to other languages as necessary.

**Comment by ejona86** [comment]
@elandau, thanks, that's helpful. I had thought about the ordering concerns of the interceptors before but forgot to mention it.

So the only solution today seems to be Context, but custom options in CallOptions would appear to be a cleaner solution. It would be good to check what the pattern would be in other languages.

**Comment by elandau** [comment]
@ejona86 Should I attempt to add this functionality to CallOptions and submit a PR?  Or would you like to wait for input on other languages before even considering this?

**Comment by ejona86** [comment]
@elandau, I still think I need a bit more buy-in on my side.

**Comment by elandau** [comment]
@ejona86 No problem.  We'll use Context for now.  Thanks for looking into this.

**Comment by buchgr** [comment]
Instead of adding custom `CallOptions` it might be cleaner to add the functionality of passing state to the `ClientInterceptor` interface itself, as this seems really only useful for interceptors?

**Comment by ejona86** [comment]
@buchgr, if we plumbed the functionality into `ClientInterceptor` interface itself, how would the client specify the configuration?

**Comment by buchgr** [comment]
Admittedly I haven't thought deeply about it, but here is my (rough) suggestion:

```java
clientStub
  .withInterceptorKey(ResponseCacheInterceptor.CACHE_KEY, "foo")
  .sayHello();
```

Introduce a `ConfigurableClientInterceptor`:

```
abstract class ConfigurableClientInterceptor extends ClientInterceptor {

  final interceptCall(MethodDescriptor, CallOptions, Channel) {
    interceptCall(MethodDescriptor, CallOptions, Channel, null);
  }

  // The same ConfObject is passed along the interceptor chain. Each call is free
  // to modify it.
  abstract interceptCall(MethodDescriptor, CallOptions, Channel, @Nullable ConfObject);
}
```

In the `InterceptorChannel`, if there is a `ConfObject` check for `ConfigurableClientInterceptor` and if true call the specific `interceptCall` method. This also allows for the mixing of `ConfigurableClientInterceptor` and `ClientInterceptor` objects.

That should work, no?

**Comment by elandau** [comment]
@buchgr Any particular reason why we can't add this to CallOptions, as I proposed in the original post?  It's already being passed into interceptCall and can easily be modified to take custom parameters as opposed to introducing a new API (ConfigurableClientInterceptor).

**Comment by buchgr** [comment]
@elandau I am not strongly against `CallOptions`, if others think it's a good idea. It just seems to me that this is specific to interceptors i.e. custom `CallOptions` don't seem useful besides your interceptor use case. Maybe it's simpler though. 

I am just throwing in ideas, so we can arrive at the best solution :-).

**Comment by ejona86** [comment]
I talked with all the languages. This seems feasible for all, except sort of iffy for C++. Depending on how interceptors are done in C++ it may be a bit unclear. Even though the other languages don't have interceptors, it seems like it would be easy with whatever the interceptor API turned out to be. Go, the only other language (currently) with interceptors, would just use Context but Context is easier to use in Go than in Java since it is passed as a parameter (not a ThreadLocal).

I've talked with some others and I was able to convince them that Context/ThreadLocal doesn't play well for this use case. So I feel comfortable moving forward with this.

There are two things I should mention about using Context though:
 1. We could have `stub.withContext()`. This would make it _much_ easier to "install" Context, but makes me very uncomfortable because it becomes hard to know which Context will be used, since a stub may have been configured `withContext()` before being passed into the current code.
 2. Even with easy Context installation, Context propagates to callbacks. So if you do an RPC on a callback, the configuration may "leak" accidentally to that other RPC. Go doesn't suffer from this because it doesn't have callbacks.
